HW1/P5/P5.py

```diff
@@ -0,0 +1,127 @@
+import pyspark
+
+# num partitions (4 times the number of cores)
+N = 256
+
+# works a lot like bfs from part 4, but instead of keeping distance as value,
+# 	it instead keeps the path that led to this node
+#  	stops when there is some path that ends with the goal (all such shortest paths
+# 	will be found in the same iteration) and returns those paths.
+def shortest_paths(start, goal, graph):
+	visited = sc.parallelize([(start, 0)], N)
+	queue = sc.parallelize([(start, (start,))], N)
+	paths = sc.emptyRDD()
+	i = 0
+	while paths.isEmpty():
+		i += 1
+		print i
+		temp = queue
+		joined = graph.join(queue, N).distinct()
+		queue = joined.values().distinct().partitionBy(N)
+		graph = graph.subtractByKey(temp, N).cache()
+		queue = queue.map(lambda (u,v): (u, v+(u,)), preservesPartitioning=True)
+		queue = queue.subtractByKey(visited, N).cache()
+		visited = visited.union(queue).partitionBy(N).cache()
+		paths = queue.values().filter(lambda x: x[-1] == goal).cache()
+	return paths.collect()
+
+# works a lot like BFS from part 4 but wraps it in a while loop until the graph
+# 	is empty of nodes (the graph is losing all nodes that have been touched at every iteration)
+def connected_components(root, graph):
+	# initialize queue and distances rdds
+	components = 0
+	biggest_component = 0
+	while not graph.isEmpty():
+		components += 1
+		print components
+		if components != 1:
+			root = graph.first()[0]
+		distances = queue = sc.parallelize([(root, 0)], N)
+		acc = sc.accumulator(1)
+		iteration = 0
+		while acc.value != 0:
+			acc = sc.accumulator(0)
+			iteration += 1
+			# keep a temporary copy of the queue to later on remove items in queue from graph
+			temp = queue
+			# get all (neighbor, distance) value pairs for each key in queue
+			joined = graph.join(queue, N).distinct()
+			# have to partition again when taking the values
+			queue = joined.values().distinct().partitionBy(N)
+			# take out original queue keys from graph and cache for later iterations
+			graph = graph.subtractByKey(temp, N).cache()
+			# set the distance values to current iteration value
+			newDistances = queue.mapValues(lambda _: iteration)
+			# remove already visited keys from the queue, and cache for later iterations
+			queue = newDistances.subtractByKey(distances, N).cache()
+			# add keys in queue with their distances to distances rdd, and cache for later iterations
+			distances = distances.union(queue).partitionBy(N).cache()
+			# increment accumulator to satisfy the while condition (unless the queue is empty, in which case the loop stops after this)
+			queue.foreach(lambda _: acc.add(1))
+		size = distances.count()
+		print size
+		biggest_component = max(biggest_component, size)
+	return components, biggest_component
```

Comment:
Didn't get to do the connected components portion, so it's interesting how it's very closely related to your BFS